### DISCUSSION TITLE
In the case of 3D geolocation data, does it account for the accuracy of location points as well?

### DISCUSSION THREAD

**Discussion by sepsol** [piercus/kalman-filter#68] [Q&A]
I couldn't find the answer by skimming through the library.

**Comment by piercus** [comment]
Not sure i get your question well, are you able to share some material ? 
Can you give an example input and the expected result ?

By accuracy, what do you mean, are you talking about the accuracy ok KM prediction/correction ? or the accuracy of the input observation ?

By "count" what do you mean exaclty ?

**Reply by sepsol** [reply]
I'm sorry for the confusion, I'm very new to the vector normalization concepts such as Kalman Filter. Let me start by explaining my use case.

I have a mobile application that is collecting geolocation data using the [@react-native-community/geolocation](https://www.npmjs.com/package/@react-native-community/geolocation) library in this format:

```json
{
  "coords": {
    "accuracy": 20,
    "altitude": 23.69999885559082,
    "heading": 0,
    "latitude": 36.5648263,
    "longitude": 53.037452,
    "speed": 0
  },
  "extras": {
    "networkLocationType": "wifi"
  },
  "mocked": false,
  "timestamp": 1702398660427
}
```

_By `accuracy` I was referring to the key that exists in the above object, I assumed that it might be a standard part of geolocation data._

The job I'm trying to accomplish is to log live user locations and show them on the map. I'm also trying to calculate the distance the user has traveled in a hilly terrain like San Francisco.

The issue I'm facing is that when the user suddenly stops moving, there is GPS jitter. I'm trying to figure out a way to reduce these inaccuracies. These inaccuracies affect how the route is displayed on the map (2D data) as well as how the distance traveled will be calculated (3D data).

<details>
  <summary><b>Screenshot of the GPS Jitter</b></summary>

  ![image](https://github.com/piercus/kalman-filter/assets/47871887/11e1e54b-352a-4037-ad7c-79d7499f71bf)

</details>

I did a little bit of [research](https://stackoverflow.com/questions/1134579/smooth-gps-data) and came across Kalman Filter and thought it would help with my issue by normalizing the logged points so that I don't have the unnecessary points causing jitter.

P.S. By 'count for' I just meant 'account for',  ESL mistake :)

**Reply by piercus** [reply]
Yes, it's possible to use Kalman-Filter to smooth your jittering. It won't do magic but it can improve.

Yes you also can use your "accuracy" parameter in the kalman filter, Is this accuracy is changing when the GPS is jittering ? or is this a constant value ? (If it's constant it won't be of any help)

I would suggest that you try to read and understand following examples : 
A. There is an example of 2D GPS localisation in https://codepen.io/piercus/pen/wvoqPww (google maps related issue, but the KF part is still working)
B. There is also an example of varying covariance for 1D https://github.com/piercus/kalman-filter/issues/34 that might be a first step to work on "accuracy" parameter

I feel you will need to mix those 2 examples, if you can start to reproduce your issue with your data based on A. i will then help you to implement the varying covariance part and link it to the "accuracy" parameter

**Reply by sepsol** [reply]
I appreciate it! Let me play around with it a bit and I'll come back here if I have any questions. Cheers 🙌🏻

**Reply by valokolaee** [reply]
Hi
My name is Mehdi 
The app is in react native

The fact is we are recording user’s location every 1 second (could be 5 sec; more or less); 
Its almost fine when it’s moving; but when it’s not the jitter happens;
The most important result we want is how to know that these moves are just jitter so that we skip theme and NOT save them 


Would you kindly please tell if there is a way, we can do that and if the library can do something to help; and if so how;


I checked the links you sent but could not understand how to use it;

We can also provide a list of coordinates to give to the library; but still don’t know which functions and how to use it

**Reply by piercus** [reply]
Hello Mehdi,

I'm concerned you did not understood how to use the exemples, can you state more precisely what is not clear ?

I would suggest you start a codepen, try to rework the example, display in a X/y graphic the input and the output, and ask me questions based on this first draft.

Alternatively you can use https://observablehq.com you will find some examples using observablehq in the README

Without more precise questions nor data, I cannot help much.

**Reply by valokolaee** [reply]
thanks

I'll give it shot

**Comment by valokolaee** [comment]
![jitter](https://github.com/piercus/kalman-filter/assets/58385770/8ab1c330-8f2f-41f1-bdf9-2b466bf4bdee)

**Comment by valokolaee** [comment]
[Uploading jiter info.zip…]()

**Reply by valokolaee** [reply]
Hereby I am providing info of how we collect data in the app,
as you we are tracking users locations in significant time period and it's all fine when user is moving;

but when the is stopped and still the geolocation tracks that are obviously NOT true and real

how can we distinguish if the movment is real or just a jitter

is there any way can use the library for this purpos and if yes how is that?

**Reply by piercus** [reply]
Hello Mehdi,

I'm concerned you did not understood how to use the exemples, can you state more precisely what is not clear ?

I would suggest you start a codepen, try to rework the example, display in a X/y graphic the input and the output, and ask me questions based on this first draft.

Alternatively you can use https://observablehq.com/ you will find some examples using observablehq in the README

Without more precise questions nor data, I cannot help much.

**Reply by valokolaee** [reply]
I need to make clear that the file coord log is having the entire log and the summary one is the of the objects that have changed in either longitude; latitude or accuracy;
